src/main/java/com/wanderersoftherift/wotr/init/WotrAttachments.java

```diff
@@ -56,12 +64,34 @@ public class WotrAttachments {
                             .build());
     public static final Supplier<AttachmentType<AbilitySlots>> ABILITY_SLOTS = ATTACHMENT_TYPES.register(
             "ability_slots",
-            () -> AttachmentType.builder(AbilitySlots::new).serialize(AbilitySlots.CODEC).copyOnDeath().build());
+            () -> AttachmentType.builder((holder) -> new AbilitySlots(holder, null))
+                    .serialize(new AttachmentSerializerFromDataCodec<AbilitySlots.Data, AbilitySlots>(
```

Comment:
To help encapsulate the data record (which I don't really feel should be exposed outside of the Attachment class) could have a static method or constant in AbilitySlots that supplies this as an IAttachmentSerializer. Then both the `data` method and record could be private.